lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -346,7 +361,7 @@ def continue_value(path, value, parent_type, field, is_non_null, ast_node, resul
           when nil
             if is_non_null
               err = parent_type::InvalidNullError.new(parent_type, field, value)
-              if !selection_result.graphql_dead
+              if !dead_result?(selection_result)
                 schema.type_error(err, context)
                 set_result(selection_result, result_name, nil)
```

Comment:
this case (and maybe others?) know they're heading into the invalid-null flow of `set_result`, I wonder if they could call it directly, and avoid the overhead of the check in `set_result` in all the other flows?

Reply:
I'm with you that the code is trying to tell us something... I tried to probe it a bit in 12fc019bf. What I found is: 

- There are basically two kinds of invalid `nil`s: 
    1. User code returned `nil` to GraphQL for a field that has `null: false`. We should warn the developer somehow (namely, the `schema.type_error` hook) because this is a bug in the system that should be fixed. 
    2. The field raise or returned a `GraphQL::ExecutionError`, so the runtime writes `nil` into the response. We assume that this error was raised on purpose and the developer doesn't need to be told, so we don't call `schema.type_error`. 
- Only the runtime code can distinguish between those, since it has the `field` definition. `set_result` has only the response data structures. (This could be modified, but that's how it is now.) That's why there's the `is_non_null` variable in `continue_value`. 
- `set_result` has to be prepared for receiving invalid nulls because error handlers don't bother tracking that. Maybe they _could_? With a method like `propagate_null`, `continue_value` could call that whenever it detects an invalid null. But even in that case, in order to propagate, we'd have to retain the result-level bookkeeping, in order to know whether a `parent` also has a non-null restriction. 

So I did a tiny hack that is _technically_ DRYer, and a hair faster, probably, since there's less work, but 🤷  not sure where to go next.